File: backend/src/main/java/com/software/contract_system/entity/ContractChange.java
package com.software.contract_system.entity;

import com.baomidou.mybatisplus.annotation.*;
import com.baomidou.mybatisplus.extension.handlers.Fastjson2TypeHandler;
import lombok.Data;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.Map;

/**
 * 合同变更申请实体
 * 对应表 t_contract_change
 * 
 * 变更类型：金额变更、时间变更、技术方案变更、附件补充、联系人变更、其他
 * 变更原因：本方主动、应乙方要求、不可抗力、政策调整、其他
 */
@Data
@TableName(value = "t_contract_change", autoResultMap = true)
public class ContractChange {

    @TableId(type = IdType.AUTO)
    private Long id;

    private Long contractId;  // 原合同ID
    private String changeNo;  // 变更单号
    private String title;     // 标题
    
    // 变更版本号 (v2.0, v3.0...)
    private String changeVersion;

    // 变更详情
    // 变更类型: AMOUNT(金额), TIME(时间), TECH(技术方案), ATTACHMENT(附件), CONTACT(联系人), OTHER(其他)
    private String changeType;
    // 变更原因: ACTIVE(本方主动), PASSIVE(应乙方要求), FORCE_MAJEURE(不可抗力), POLICY(政策调整), OTHER(其他)
    private String reasonType;
    private String description;  // 变更说明
    
    // 乙方沟通记录
    private String partyBCommunication;

    // ★★★ 核心：变更前后的对比数据 (JSON) ★★★
    // 包含 beforeContent 和 afterContent
    @TableField(typeHandler = Fastjson2TypeHandler.class)
    private Map<String, Object> diffData;
    
    // 是否重大变更（金额>20%或核心条款变更）
    private Boolean isMajorChange;
    
    // 变更金额差额
    private BigDecimal amountDiff;

    // 状态: 0:草稿, 1:审批中, 2:已通过, 3:已驳回, 4:已撤销
    private Integer status;
    
    // 审批流程实例ID
    private Long instanceId;

    private Long initiatorId; // 发起人

    @TableField(fill = FieldFill.INSERT)
    private LocalDateTime createdAt;
    
    // 审批完成时间
    private LocalDateTime approvedAt;
    
    // 变更生效时间
    private LocalDateTime effectiveAt;
    
    // 附件路径（多个用逗号分隔）
    private String attachmentPath;
}

File: frontend/src/api/contractChange.ts
import request from '@/utils/request'

// 变更类型选项
export const CHANGE_TYPE_OPTIONS = [
  { value: 'AMOUNT', label: '金额变更' },
  { value: 'TIME', label: '时间变更' },
  { value: 'TECH', label: '技术方案变更' },
  { value: 'ATTACHMENT', label: '附件补充' },
  { value: 'CONTACT', label: '联系人变更' },
  { value: 'OTHER', label: '其他' }
]

// 变更原因选项
export const REASON_TYPE_OPTIONS = [
  { value: 'ACTIVE', label: '本方主动' },
  { value: 'PASSIVE', label: '应乙方要求' },
  { value: 'FORCE_MAJEURE', label: '不可抗力' },
  { value: 'POLICY', label: '政策调整' },
  { value: 'OTHER', label: '其他' }
]

// 变更状态选项
export const CHANGE_STATUS_OPTIONS = [
  { value: 0, label: '草稿', type: 'info' },
  { value: 1, label: '审批中', type: 'warning' },
  { value: 2, label: '已通过', type: 'success' },
  { value: 3, label: '已驳回', type: 'danger' },
  { value: 4, label: '已撤销', type: 'info' }
]

export interface ContractChangeDTO {
  contractId: number
  title: string
  changeType: string
  reasonType: string
  description: string
  partyBCommunication?: string
  newName?: string
  newAmount?: number
  newContent?: string
  newPartyB?: string
  newAttributes?: Record<string, any>
  attachmentPath?: string
}

export interface ContractChangeVO {
  id: number
  contractId: number
  contractNo: string
  contractName: string
  changeNo: string
  title: string
  changeVersion: string
  changeType: string
  changeTypeName: string
  reasonType: string
  reasonTypeName: string
  description: string
  partyBCommunication?: string
  diffData?: Record<string, any>
  diffItems?: DiffItem[]
  isMajorChange: boolean
  amountDiff?: number
  changePercent?: number
  status: number
  statusName: string
  instanceId?: number
  initiatorId: number
  initiatorName: string
  createdAt: string
  approvedAt?: string
  effectiveAt?: string
  attachmentPath?: string
}

export interface DiffItem {
  fieldName: string
  fieldLabel: string
  beforeValue: any
  afterValue: any
  changeDesc: string
}

/**
 * 创建变更申请（草稿）
 */
export function createChange(data: ContractChangeDTO) {
  return request({
    url: '/contract-change/create',
    method: 'post',
    data
  })
}

/**
 * 提交变更审批
 */
export function submitChange(changeId: number) {
  return request({
    url: `/contract-change/${changeId}/submit`,
    method: 'post'
  })
}

/**
 * 撤销变更申请
 */
export function cancelChange(changeId: number) {
  return request({
    url: `/contract-change/${changeId}/cancel`,
    method: 'post'
  })
}

/**
 * 获取变更详情
 */
export function getChangeDetail(changeId: number) {
  return request<ContractChangeVO>({
    url: `/contract-change/${changeId}`,
    method: 'get'
  })
}

/**
 * 获取合同的变更历史
 */
export function getChangeHistory(contractId: number) {
  return request<ContractChangeVO[]>({
    url: `/contract-change/history/${contractId}`,
    method: 'get'
  })
}

/**
 * 分页查询变更申请
 */
export function listChanges(params: {
  pageNum?: number
  pageSize?: number
  contractId?: number
  status?: number
  initiatorId?: number
}) {
  return request({
    url: '/contract-change/list',
    method: 'get',
    params
  })
}

/**
 * 我的变更申请
 */
export function myChanges(params: {
  pageNum?: number
  pageSize?: number
  status?: number
}) {
  return request({
    url: '/contract-change/my',
    method: 'get',
    params
  })
}

/**
 * 检查合同是否可以发起变更
 */
export function canCreateChange(contractId: number) {
  return request<{ canChange: boolean; message: string }>({
    url: `/contract-change/can-change/${contractId}`,
    method: 'get'
  })
}

/**
 * 重大变更预检
 */
export function checkMajorChange(data: ContractChangeDTO) {
  return request<{ isMajorChange: boolean; message: string }>({
    url: '/contract-change/check-major',
    method: 'post',
    data
  })
}

/**
 * 获取当前用户可变更的合同列表
 * 返回用户有权限变更的已生效合同（创建者或审批参与者）
 */
export function getChangeableContracts() {
  return request<any[]>({
    url: '/contract-change/changeable-contracts',
    method: 'get'
  })
}

